plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/editor/CodeWhispererEditorManagerNew.kt

```diff
@@ -0,0 +1,282 @@
+        val typeahead = preview.typeahead
+        val detail = preview.detail
+        val userInput = preview.userInput
+        val reformatted = CodeWhispererPopupManagerNew.getInstance().getReformattedRecommendation(
+            detail,
+            userInput
+        )
+        val remainingRecommendation = reformatted.substring(typeahead.length)
+        val originalOffset = primaryCaret.offset - typeahead.length
+
+        val endOffset = primaryCaret.offset + remainingRecommendation.length
+
+        val insertEndOffset = sessionContext.insertEndOffset
+        val endOffsetToReplace = if (insertEndOffset != -1) insertEndOffset else primaryCaret.offset
+
+        preview.detail.isAccepted = true
+
+        WriteCommandAction.runWriteCommandAction(project) {
+            document.replaceString(originalOffset, endOffsetToReplace, reformatted)
+            PsiDocumentManager.getInstance(project).commitDocument(document)
+            primaryCaret.moveToOffset(endOffset + detail.rightOverlap.length)
+        }
+
+        ApplicationManager.getApplication().invokeLater {
+            WriteCommandAction.runWriteCommandAction(project) {
+                val rangeMarker = document.createRangeMarker(originalOffset, endOffset, true)
+
+                CodeWhispererTelemetryServiceNew.getInstance().enqueueAcceptedSuggestionEntry(
+                    detail.requestId,
+                    requestContext,
+                    responseContext,
+                    Instant.now(),
+                    PsiDocumentManager.getInstance(project).getPsiFile(document)?.virtualFile,
+                    rangeMarker,
+                    remainingRecommendation,
+                    selectedIndex,
+                    detail.completionType
+                )
+
+                ApplicationManager.getApplication().messageBus.syncPublisher(
+                    CodeWhispererPopupManager.CODEWHISPERER_USER_ACTION_PERFORMED,
+                ).afterAccept(states, previews, sessionContext, rangeMarker)
+            }
+        }
+    }
+
+    private fun isMatchingSymbol(symbol: Char): Boolean =
+        PAIRED_BRACKETS.containsKey(symbol) || PAIRED_BRACKETS.containsValue(symbol) || PAIRED_QUOTES.contains(symbol) ||
+            symbol.isWhitespace()
+
+    fun getUserInputSinceInvocation(editor: Editor, invocationOffset: Int): String {
+        val currentOffset = editor.caretModel.primaryCaret.offset
+        return editor.document.getText(TextRange(invocationOffset, currentOffset))
+    }
+
+    fun getCaretMovement(editor: Editor, caretPosition: CaretPosition): CaretMovement {
+        val oldOffset = caretPosition.offset
+        val newOffset = editor.caretModel.primaryCaret.offset
+        return when {
+            oldOffset < newOffset -> CaretMovement.MOVE_FORWARD
+            oldOffset > newOffset -> CaretMovement.MOVE_BACKWARD
+            else -> CaretMovement.NO_CHANGE
+        }
+    }
+
+    fun getMatchingSymbolsFromRecommendation(
```

Comment:
if it's not too hard can you pull out the common code so there's less delta and risk